src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php

```diff
@@ -484,7 +484,15 @@ public function firstOrCreate(array $attributes, array $joining = [], $touch = t
      */
     public function updateOrCreate(array $attributes, array $values = [], array $joining = [], $touch = true)
     {
-        if (is_null($instance = $this->where($attributes)->first())) {
+        $instance = null;
+
+        // Use the models that have already been loaded by the parent. This will solve
+        // the n + 1 query problem for the developer and also increase performance.
+        if (empty($attributes)) {
+            $instance = $this->parent->getRelation($this->related->getTable())->first();
```

Comment:
The relationship isn't always named like the table, we don't know the actual name. If it's different, you get an exception.